# Exercise Walk-through

## 01 Import Libraries

In [11]:
import pandas as pd
import numpy as np
import os

## 02 Import Data Sets

In [15]:
path=r"C:\Users\sahin\Documents\Instacart Basket Analysis"

In [ ]:
# Import Orders Data

In [16]:
df_ords = pd.read_csv(os.path.join(path, "02 Data", "Original Data", "orders.csv"), index_col = False)

In [ ]:
# Import Product Data

In [17]:
df_prods = pd.read_csv(os.path.join(path, "02 Data", "Original Data", "products.csv"), index_col = False)

## 03 Data Wrangling Procedures

In [ ]:
# Drop the eval_set column from orders 

In [18]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [ ]:
# Count NaN values in the days_since_prior_order column

In [9]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [ ]:
# Rename the order_dow column into orders_day_of_week

In [20]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_the_week'}, inplace = True)

In [21]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2398795        1    prior             2                       3   
2    473747        1    prior             3                       3   
3   2254736        1    prior             4                       4   
4    431534        1    prior             5                       4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [ ]:
# Change the data type of the order_id column

In [23]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [ ]:
# check the data type of the order_id column

In [34]:
df_ords['order_id'].dtype

dtype('O')

In [ ]:
# Import Department Data

In [39]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [40]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [ ]:
# Transpose the table

In [44]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [ ]:
# Creating a new dataframe

In [45]:
df_dep_t = df_dep.T

In [46]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [ ]:
# Adding an index to dataframe

In [47]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [ ]:
# Take the first row of df_dep_t for the header

In [58]:
new_header = df_dep_t.iloc[0]

In [59]:
new_header

0    department
Name: department_id, dtype: object

In [ ]:
# Delete the first row

In [60]:
df_dep_t_new = df_dep_t[1:]

In [61]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [ ]:
# Set a new header

In [63]:
df_dep_t_new.columns = new_header

In [64]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 04 Data Dictionaries

In [ ]:
# Create a data dictionary

In [67]:
data_dict = df_dep_t_new.to_dict('index')

In [68]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [ ]:
# Use the data dictionary in practice

In [69]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [70]:
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting

In [ ]:
# Create a dataframe that only contains data from the snacks department

In [72]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [73]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [ ]:
# Saving the results into a new dataframe df_snacks

In [74]:
df_snacks = df_prods[df_prods['department_id']==19]

In [75]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [ ]:
The second way is the Loc function looks for a column name

In [76]:
df_snacks_2 = df_prods.loc[df_prods['department_id']==19]

In [77]:
df_snacks_2

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [ ]:
# Using a list is handy if you want to check multiple departments at once

In [82]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

In [85]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([17,18,19])]

# Directions 2-14

In [ ]:
# 2.Find another identifier variable in the df_ords dataframe that doesn't need to be included in your 
#   analysis as a numeric variable and change it to a suitable format.

In [90]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0  2539329        1    prior             1                       2   
1  2398795        1    prior             2                       3   
2   473747        1    prior             3                       3   
3  2254736        1    prior             4                       4   
4   431534        1    prior             5                       4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [93]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [94]:
df_ords['user_id'].dtype

dtype('O')

In [ ]:
# 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without
##   overwriting the dataframe.

In [100]:
df_new = df_ords.rename(columns = {'order_hour_of_day' : 'order_time'}, inplace = False)

In [101]:
df_new.head()

order_id user_id eval_set  order_number  orders_day_of_the_week  order_time  \
0  2539329       1    prior             1                       2           8   
1  2398795       1    prior             2                       3           7   
2   473747       1    prior             3                       3          12   
3  2254736       1    prior             4                       4           7   
4   431534       1    prior             5                       4          15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [ ]:
# 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of 
#    the corresponding variable and share your findings.

In [102]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

The busiest hour of the day is 10 am. 288418 orders have been placed.

In [ ]:
# 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods 
#    dataframe using a data dictionary.

In [103]:
print(data_dict.get('4'))

{'department': 'produce'}


The department_id 4 is the produce department.

In [ ]:
# 6. The sales team in your client’s organization wants to know more about breakfast item sales. 
#    Create a subset containing only the required information.

In [106]:
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [108]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [ ]:
# 7. They’d also like to see details about customers who might be throwing dinner parties. 
#    Your task is to find all observations from the entire dataframe that include items from 
#    the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present 
#    this subset to your client.

In [109]:
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [110]:
df_dinner_party.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

In [ ]:
# 8. It’s important that you keep track of total counts in your dataframes. How many rows does 
#    the last dataframe you created have?

In [114]:
df_dinner_party.shape

(7650, 5)

7650 rows

In [ ]:
# 9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange 
#    about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [115]:
df_user_1 = df_ords[df_ords['user_id']=='1']

In [116]:
df_user_1.head(30)

order_id user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329       1    prior             1                       2   
1   2398795       1    prior             2                       3   
2    473747       1    prior             3                       3   
3   2254736       1    prior             4                       4   
4    431534       1    prior             5                       4   
5   3367565       1    prior             6                       2   
6    550135       1    prior             7                       1   
7   3108588       1    prior             8                       1   
8   2295261       1    prior             9                       1   
9   2550362       1    prior            10                       4   
10  1187899       1    train            11                       4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

In [ ]:
# 10. You also need to provide some details about this user’s behavior. What basic stats can you 
#     provide based on the information you have?
    

In [117]:
df_user_1.describe()

order_number  orders_day_of_the_week  order_hour_of_day  \
count     11.000000               11.000000          11.000000   
mean       6.000000                2.636364          10.090909   
std        3.316625                1.286291           3.477198   
min        1.000000                1.000000           7.000000   
25%        3.500000                1.500000           7.500000   
50%        6.000000                3.000000           8.000000   
75%        8.500000                4.000000          13.000000   
max       11.000000                4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [ ]:
# 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [119]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [ ]:
# 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder 
#     so that you have a “.csv” file of your departments data in the correct format.

In [120]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))